linkerd/app/src/trace_collector.rs

```diff
@@ -0,0 +1,120 @@
+use linkerd_app_core::http_tracing::{CollectorProtocol, SpanSink};
+use linkerd_app_core::metrics::ControlHttp as HttpMetrics;
+use linkerd_app_core::svc::NewService;
+use linkerd_app_core::{control, dns, identity, opencensus, opentelemetry};
+use linkerd_error::Error;
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use tracing::info_span;
+
+pub mod oc_collector;
+pub mod otel_collector;
+
+#[derive(Clone, Debug)]
+pub enum Config {
+    Disabled,
+    Enabled(Box<EnabledConfig>),
+}
+
+#[derive(Clone, Debug)]
+pub struct EnabledConfig {
+    pub control: control::Config,
+    pub attributes: HashMap<String, String>,
+    pub hostname: Option<String>,
+    pub kind: CollectorProtocol,
+}
+
+pub type Task = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+pub enum TraceCollector {
+    Disabled,
+    Enabled(Box<EnabledCollector>),
+}
+
+pub struct EnabledCollector {
+    pub addr: control::ControlAddr,
+    pub kind: CollectorProtocol,
+    pub span_sink: SpanSink,
+    pub task: Task,
+}
+
+impl EnabledCollector {
+    pub fn info_span(&self) -> tracing::Span {
+        match self.kind {
+            CollectorProtocol::OpenCensus => info_span!("opencensus"),
+            CollectorProtocol::OpenTelemetry => info_span!("opentelemetry"),
```

Comment:
This seems fishy to me... these span macros are going to capture the context in this file and not at the callsite. Do we really need to differentiate the span based on the collector type? Would it be sufficient to just use something more general like `info_span!("tracing")`?

Reply:
In this case I think a single `tracing` span sufficient, fixed.